exporter/fileexporter/grouping_file_exporter.go

```diff
@@ -0,0 +1,249 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package fileexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/fileexporter"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/golang-lru/v2/simplelru"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+)
+
+type groupingFileExporter struct {
+	logger        *zap.Logger
+	marshaller    *marshaller
+	pathPrefix    string
+	pathSuffix    string
+	attribute     string
+	maxOpenFiles  int
+	newFileWriter func(path string) (*fileWriter, error)
+
+	mutex   sync.Mutex
+	writers *simplelru.LRU[string, *fileWriter]
+}
+
+func (e *groupingFileExporter) consumeTraces(ctx context.Context, td ptrace.Traces) error {
+	if td.ResourceSpans().Len() == 0 {
+		return nil
+	}
+
+	groups := make(map[string][]ptrace.ResourceSpans)
+
+	for i := 0; i < td.ResourceSpans().Len(); i++ {
+		rSpans := td.ResourceSpans().At(i)
+		group(e, groups, rSpans.Resource(), rSpans)
+	}
+
+	var errs error
+	for pathSegment, rSpansSlice := range groups {
+		traces := ptrace.NewTraces()
+		for _, rSpans := range rSpansSlice {
+			rSpans.CopyTo(traces.ResourceSpans().AppendEmpty())
+		}
+
+		buf, err := e.marshaller.marshalTraces(traces)
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+
+		err = e.write(ctx, pathSegment, buf)
+		if err != nil {
+			errs = errors.Join(errs, err)
+		}
+	}
+
+	return errs
```

Comment:
The most likely errors here are permission errors when creating directories or writing to files.

What's the result of returning this error from the exporter? Is the error propagated back, or is it retried? Is there any way of marking this error as non-retryable, so the collector simply logs it as an error and moves on?

Reply:
Good question. Maybe @dmitryax knows?

Reply:
Here's my understanding, correct me if I'm wrong.

The file exporter seems to not be using the backoff/retry or queueing functionality from exporterhelper, so this error will be simply returned to the previous component in the pipeline (a processor or a receiver). The error is not wrapped into a [permanent](https://github.com/open-telemetry/opentelemetry-collector/blob/v0.95.0/consumer/consumererror/permanent.go) error with NewPermanent(errs), so the preceding component might or might not retry sending the data, depending on its configuration and/or implementation.

I suppose that's the right way to do. If you are certain the error is permanent, you might want to wrap it with `NewPermanent`.

Reply:
Thanks @astencel-sumo 

I did a quick search in the code base for "errors.Join", to find any other instances where a component might have partial errors (eg: some logs are processed properly, others are not).

I didn't find anything indicating that these partial errors are handled (eg: only the failed logs would be retried).

In some exporters (eg: awscloudwatchlogsexporter) I see it uses retry/queue using the erporterhelper that you mentioned. In some cases I think that might result in duplicate logs. The chance of a partial failure might be slim with awscloudwatchlogsexporter, but it is much higher with this fileexporter.

Anyway, considering most other exporters/connectors simply return these joined errors as it is here, I would leave it as it is. And I think adding queue/backoff/retry capabilities is outside the scope of this pr.

So, what do you think about merging this as it is, @astencel-sumo , @djaglowski ?

Reply:
Retriable errors can embed the portion of the data that can be retried to avoid double publishing. But, in this case I believe the error is permanent, so just wrap it in https://github.com/open-telemetry/opentelemetry-collector/blob/fb912b97c0e98b380f9ebf9e70e81ece23e3fdb2/consumer/consumererror/permanent.go#L16C6-L16C18